Sources/swift-format/main.swift

```diff
@@ -16,58 +16,46 @@ import SwiftFormatConfiguration
 import SwiftFormatCore
 import SwiftSyntax
 import TSCBasic
-import TSCUtility
 
-fileprivate func main(_ arguments: [String]) -> Int32 {
-  let url = URL(fileURLWithPath: arguments.first!)
-  let options = processArguments(commandName: url.lastPathComponent, Array(arguments.dropFirst()))
-  let diagnosticEngine = makeDiagnosticEngine()
-  switch options.mode {
-  case .format:
-    if options.paths.isEmpty {
-      let configuration = loadConfiguration(
-        forSwiftFile: nil, configFilePath: options.configurationPath)
-      return Int32(
-        formatMain(
+extension SwiftFormatCommand {
+  func run() throws {
```

Comment:
Since `ArgumentParser` has subcommand support, it would make sense to have `format`, `lint`, and `dump-configuration` be subcommands with `format` as the default subcommand. It looks like that would also improve the command line validation because options could be restricted to just the commands that they apply to. **But** we also have internal tooling that depends on the current command line interface, so we wouldn't want to do that breaking change here yet.

As a possible follow-up PR, does `ArgumentParser`'s design let us easily create those subcommands and then have the old `-m/--mode` flag delegate to them as a migration path so that we could have both interfaces implemented simultaneously, and then once we have our other tooling no longer depending on the old interface, we could remove it from `swift-format`?

Reply:
We should totally be able to have those parallel interfaces. I have an old version of `swift-format` migrated to a subcommand interface, so after this is ready to merge I'll update that and open another PR.